subprojects/core-api/src/main/java/org/gradle/api/file/SourceDirectorySet.java

```diff
@@ -127,4 +128,19 @@
      * @since 4.0
      */
     void setOutputDir(File outputDir);
+
+    /**
+     * Returns the directory property that is, or can be, bound to the task that produces the output.
+     * Use this as part of a classpath or input to another task to ensure that the output is created before it is used.
+     *
+     * Set this property to a output property of a task to connect this source directory set to the output producing (compile) task.
+     *
+     * Note: To define the path of the output folder for a source directory set that is already wired to
+     * a task (e.g. 'java' / `compileJava`) use {@link #setOutputDir(Provider)} or {@link #setOutputDir(File)}.
+     *
+     * @return The output directory property for this set of sources.
+     * @since 6.1
+     */
+    @Incubating
+    DirectoryProperty getOutputDirectoryProperty();
```

Comment:
This is a bit problematic, because it would be easy for someone to assume that they can set this to a directory and have the output produced there. However, doing this would instead silently break the connection to the compile task and everything would be broken.

An alternative approach would be to have this method return `Provider<Directory>`, to reflect the fact this this value is intended to only be _read_, and add a method which takes care of wiring the output property of the producing task so that the task produces its output into the correct location and so that the value returned from `getOutputDirectory()` can be used as an input for consuming tasks (the property needs a name that better reflects its role here, eg `getClassesDirectory()` or something). Something like this:

```java
interface SourceDirectorySet {
    // Allows build logic to configure where the outputs should end up
    DirectoryProperty getDestinationDirectory() // replaces the existing outputDir property

    // Allow build logic to consume the source set output
    Provider<Directory> getClassesDirectory()

    // registers and connects the task that produces the output of this source set
    <T extends Task> void compiledBy(TaskProvider<T> taskProvider, Function<T, DirectoryProperty> mapping)

    // could also have a convenience for AbstractCompileTask
    void compiledBy(TaskProvider<? extends AbstractCompileTask> taskProvider)
}

// In a plugin:
SourceDirectorySet sourceSet = ...
TaskProvider<MyCompileTask> taskProvider = ...
sourceSet.compiledBy(taskProvider, t -> { it.myOutputDirectory });

// The implementation would look something like:
class DefaultSourceDirectorySet implements SourceDirectorySet {
    private final DirectoryProperty destinationDir = ...
    private final DirectoryProperty classesDir = ...

    DirectoryProperty getDestinationDirectory() { return destinationDir; }
    Provider<Directory> getClassesDirectory() { return classesDir; }

    <T extends Task> void compiledBy(TaskProvider<T> taskProvider, Function<T, DirectoryProperty> mapping) {
        // connect the things
        taskProvider.configure(t -> { mapping(t).set(destinationDirectory) });
        classesDir = taskProvider.flatMap(mapping);
    }
}
```

It would be even better to extract this pattern to a new type. We have exactly the same issues and use cases in the native plugins, for example

Reply:
Thanks for the suggestion. I tried to avoid this at first because avoid adding compiledBy/builtBy to that interface was what triggered this whole work. But looking at it again from some distance, I agree that this makes most sense.

I also wanted to add the convenience, but `AbstractCompileTask` is not a core API type. So I left it out for now, but added it as example to the Javadoc of `compiledBy`.